CMSgov/mint-app
# [EASI-2346] - Faktory Jobs

# EASI-2346

## Changes and Description

- Add faktory jobs for daily digest
 ```mermaid
graph TD;
    DailyDigestCronJob-->AnalyzedAuditBatchJob;
    AnalyzedAuditBatchJob-->AnalyzedAuditJob1;
    AnalyzedAuditBatchJob-->AnalyzedAuditJob2;
    AnalyzedAuditBatchJob-->AnalyzedAuditJob3;
    AnalyzedAuditBatchJob-->AnalyzedAuditJobX;
    AnalyzedAuditJob1-->AnalyzedAuditBatchJobSuccess;
    AnalyzedAuditJob2-->AnalyzedAuditBatchJobSuccess;
    AnalyzedAuditJob3-->AnalyzedAuditBatchJobSuccess;
    AnalyzedAuditJobX-->AnalyzedAuditBatchJobSuccess;
    AnalyzedAuditBatchJobSuccess-->DigestEmailBatchJob;
    DigestEmailBatchJob-->DigestEmailJob1;
    DigestEmailBatchJob-->DigestEmailJob2;
    DigestEmailBatchJob-->DigestEmailJob3;
    DigestEmailBatchJob-->DigestEmailJobX;
    DigestEmailJob1-->DigestEmailBatchJobSuccess;
    DigestEmailJob2-->DigestEmailBatchJobSuccess;
    DigestEmailJob3-->DigestEmailBatchJobSuccess;
    DigestEmailJobX-->DigestEmailBatchJobSuccess;
```
- Add `Worker` field to `Server`  struct so the worker can get setup in `routes` with config it needs there then start where the server starts. I was seeing some weird behavior starting a go routine in `routes.go`. Plus I think it makes more sense to pull it out and start it where we start the server.
- Minor updates to existing code.
- Renamed files in worker to try and keep it more consistent/cleaner. Basically one file per job or batch/job combination.

### Cron Config
- TOML config files for cron jobs are in  `/.faktory` folder. We may want to move the prod and other envs to the infra repo in the future. In deployed environments this file must be loaded into `/etc/faktory/conf.d` on startup. See: [Ent Cron](https://github.com/contribsys/faktory/wiki/Ent-Cron)

### Environment Variables
**NOTE:** these have been updated in all deployed environments with `FAKTORY_PROCESS_JOBS=false`
- `FAKTORY_PROCESS_JOBS` - whether to process the jobs in the queue or not. This should default to false in dev unless you are actively developing on jobs. 

- `FAKTORY_URL` - this is the url of TCP server for the client to talk to and for the faktory_worker to process jobs. Port 7419 is the default port. 

- `FAKTORY_CONNECTIONS` -  this is the max number of connections to the faktory server (think database connection pool). Our enterprise license has a max of 100 connections that we can use in any environment (development does not count against this number) For example: `prod - 40`, `impl - 20`, `dev - 20`, `test - 20` 

## How to test this change
To run locally:
1. set `FAKTORY_PROCESS_JOBS` in `.envrc.local` to `true`
3. run backend `scripts/dev up:backend`
4. The first time you try to download `faktory-ent` image it will ask you for username and password for `docker.contribsys.com`. These credentials are in 1Password.
5. seed database
6. wait for cron job to run. it will run every 5th minute. you can modify the schedule in `.faktory/conf.d/cron.toml` if you want.
7. Wait for cron to run. Good time to make sure your database is seeded.  You can check status of jobs at `localhost:7420`
8. Once all jobs run you should see emails in the mail catcher `http://localhost:1085/` and `analyzed_audits` in the database

Run tests:
1. `APP_ENV=testing scripts/dev up:backend`
2. `scripts/dev test:go:only "github.com/cmsgov/mint-app/pkg/worker"`

## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
